collector/filefd_linux.go

```diff
@@ -0,0 +1,94 @@
+// +build !nonetstat
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+const (
+	procFileFDStat      = "/proc/sys/fs/file-nr"
+	fileFDStatSubsystem = "filefd"
+)
+
+type fileFDStatCollector struct {
+	metrics map[string]prometheus.Gauge
+}
+
+func init() {
+	Factories[fileFDStatSubsystem] = NewFileFDStatCollector
+}
+
+// NewFileFDStatCollector returns a new Collector exposing file-nr stats
+func NewFileFDStatCollector() (Collector, error) {
+	return &fileFDStatCollector{
+		metrics: map[string]prometheus.Gauge{},
+	}, nil
+}
+
+func (c *fileFDStatCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	fileNRStats, err := getFileFDStats(procFileFDStat)
+	if err != nil {
+		return fmt.Errorf("couldn't get file-nr: %s", err)
+	}
+	for protocol, protocolStats := range fileNRStats {
+		for name, value := range protocolStats {
+			key := protocol + "_" + name
+			if _, ok := c.metrics[key]; !ok {
+				c.metrics[key] = prometheus.NewGauge(
+					prometheus.GaugeOpts{
+						Namespace: Namespace,
+						Subsystem: fileFDStatSubsystem,
+						Name:      key,
+						Help:      fmt.Sprintf("%s %s from %s.", protocol, name, procFileFDStat),
+					},
+				)
+			}
+			v, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return fmt.Errorf("invalid value %s in file-nr: %s", value, err)
+			}
+			c.metrics[key].Set(v)
+		}
+	}
+	for _, m := range c.metrics {
+		m.Collect(ch)
+	}
+	return err
+}
+
+func getFileFDStats(fileName string) (map[string]map[string]string, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseFileFDStats(file, fileName)
+}
+
+func parseFileFDStats(r io.Reader, fileName string) (map[string]map[string]string, error) {
+	var (
+		sockStat = map[string]map[string]string{}
```

Comment:
This whole datastructure seems overkill for essentially returning 3 values. Also, why is the constant protocol in there needed?